oauth2_client/client.py

```diff
@@ -29,9 +38,58 @@ class OAuth2Client(OAuth2Session):
         > r = client.get('/api/license/1/detail/')
     """
 
-    def __init__(self, app, token):
-        self.service_host = app.service_host  # used to transform relative URLs to absolute
-        super(OAuth2Client, self).__init__(client_id=app.client_id, token=token)
+    """
+    If token re-fetch happens more than once in this period, raise an exception. Something is wrong,
+    don't spam the provider.
+    """
+    REAUTH_LIMIT_SECONDS = 10
+
+    def __init__(self, token, expired_callback=None):
+        """
+        Create OAuth2Client
+
+        :param token: oauth2_client.model.AccessToken object
+        :param expired_callback: (optional) function accepting one parameter, `Application`. It is
+            called when TokenExpiredError is detected or 403 Forbidden status code received.
+            It should return a valid (not expired) oauth2_client.model.AccessToken object for
+            communication with Resource Owner specified by `Application`.
+            Explanation: The parent class, OAuth2Session already implements automated token refresh,
+            but their solution works only for flows that issue a refresh_token. Both currently
+            supported flows, JWT and Client Credentials don't provide the refresh_token, so we have
+            to handle the refresh ourselves. (Client Credentials MAY issue refresh_token but it
+            doesn't in case of the `oauth2_provider` we use). If the token expiry is detected more
+            than once within `REAUTH_LIMIT_SECONDS` period, `oauthlib.oauth2.TokenExpiredError`
+            is raised (safety fuse approach).
+        """
+        self.app = token.application  # Application this client talks to
+        self.last_token_fetch_time = None  # used for the safety fuse preventing too often re-authorization
+        self.service_host = self.app.service_host  # used to transform relative URLs to absolute
+        self.expired_callback = self._install_callback(expired_callback) if expired_callback else None
+        super(OAuth2Client, self).__init__(client_id=self.app.client_id, token=token.to_client_dict())
+
+    def _install_callback(self, expired_callback):
+        """
+        Wraps provided expiry callback into a safety fuse, so we won't spam the provider with token
+        requests in case of problems with the client.
+
+        :param expired_callback: a function taking one argument, `oauth2_client.models.Application`.
+            It has to fetch and store new AccessToken.
+        :return:  a wrapped callback function
+        """
+        def wrapped_callback(app):
+            if self.last_token_fetch_time \
+                    and self.last_token_fetch_time + timedelta(seconds=self.REAUTH_LIMIT_SECONDS) >= timezone.now():
```

Comment:
This ugliness copes with the single startup case where there has been no previous fetch.

Simpler rather to initialise `self.last_token_time` to now minus one minute,* and then the `None` test can be omitted, leading to a simpler logic flow and a truer assertion of initialization conditions: "No token requests in the last 60 seconds" is better than "No token requests yet." Explicit is better than implicit.

* In fact, `now() - REAUTH_LIMT_SECONDS`

Reply:
outdated due to introduction of 3rd party implementation of a circuit breaker pattern